flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java

```diff
@@ -197,6 +208,88 @@ public void shutdown(Time timeout) {
 				executor);
 	}
 
+	public <M extends MessageHeaders<R, P, U>, U extends MessageParameters, R extends RequestBody, P extends ResponseBody> CompletableFuture<P> sendRequest(
+			String targetAddress,
+			int targetPort,
+			M messageHeaders,
+			U messageParameters,
+			R request,
+			Collection<Path> jars,
+			Collection<Path> userArtifacts) throws IOException {
```

Comment:
It would be great if the `RestClient` did not know about the distinction between `jars` and `userArtifacts`. Instead it should be enough to provide this method a collection of `FileUpload` objects which contain the path and some meta information to make sense of the different files contained in the body on the receiving side.

Reply:
So far we managed to not expose netty stuff in the `RestClient` API, I would prefer if we didn't start now.

Could we not just pass a single `Collection<Path>` instead? The content-type would then always be `application/octet-stream`.

Reply:
I didn't mean to use Netty's `FileUpload` class but creating our own where we could specify the content type and other information. But we can begin also with `Collection<Path>` and encode what kind of file it is in the file name, for example.